### DISCUSSION TITLE
Restartable programs

### DISCUSSION THREAD

**Discussion by oliverschmidt** [cc65/cc65#2501] [Ideas]
Hi,

I certainly haven't thought this through completely but...

When creating a program to be flashed to a Cart, then:
* No initialization code gets overwritten - as one can't overwrite ROM.
* The .DATA segment gets copied to RAM - as one can't write to it in ROM.

As a side effect, a Cart program is always restartable. Coming from there, the path to a disk-based restartable program seems pretty easy: Pretend that it is a Cart program and do exactly the two things listed above. In particular, copy the .DATA segment (just not from ROM to RAM but from RAM to RAM).

BTW: https://github.com/cc65/cc65/discussions/2103 is related.

**Comment by mrdudz** [comment]
Sure thing, thats how usually crt0 works on "bigger" systems. But its not really an option for most non trivial programs written with cc65 - because of memory constraints

**Comment by oliverschmidt** [comment]
My experience suggests that the .DATA segment tends to be rather small. And the initialization code space wasn't reused on most targets most of the time cc65 exists. So I disagree with your view that this is not an option for most non-trivial programs.

**Comment by groessler** [comment]
I definitely remember it being convenient on the C64 to re-run a program after exit.
@oliverschmidt, what do you suggest? An additional (linker?) config to create restartable programs?

**Comment by mrdudz** [comment]
It'd need a different crt0 (which copies DATA) and linker config (which reserves space for the copied DATA in BSS or sth)

**Comment by oliverschmidt** [comment]
> I definitely remember it being convenient on the C64 to re-run a program after exit.

I remember the discussions back then. It "sort of" worked for "a lot of" programs. Therefore I advocated back then to remove the option. But here I'm proposing a more heavy-weight approach that definitively works for all programs which can afford the memory overhead.

**Comment by mrdudz** [comment]
Seems that this is yet another case of "we need a good way to link different stuff based on options"

I still think splitting the library is the best thing we can do there, ie make it more like it works eg with gcc. The monolithic approach creates more hassle than it is worth, IMHO
